java/ql/src/experimental/Security/CWE/CWE-400/ThreadResourceAbuse.ql

```diff
@@ -0,0 +1,87 @@
+/**
+ * @name Uncontrolled thread resource consumption
+ * @description Use user input directly to control thread sleep time could lead to performance problems
+ *              or even resource exhaustion.
+ * @kind path-problem
+ * @id java/thread-resource-abuse
+ * @tags security
+ *       external/cwe/cwe-400
+ */
+
+import java
+import ThreadPauseSink
+import semmle.code.java.dataflow.FlowSources
+import DataFlow::PathGraph
+
+/** The `getInitParameter` method of servlet or JSF. */
+class GetInitParameter extends Method {
+  GetInitParameter() {
+    (
+      this.getDeclaringType()
+          .getASupertype*()
+          .hasQualifiedName(["javax.servlet", "jakarta.servlet"],
+            ["FilterConfig", "Registration", "ServletConfig", "ServletContext"]) or
+      this.getDeclaringType()
+          .getASupertype*()
+          .hasQualifiedName(["javax.faces.context", "jakarta.faces.context"], "ExternalContext")
+    ) and
+    this.getName() = "getInitParameter"
+  }
+}
+
+/** An access to the `getInitParameter` method. */
+class GetInitParameterAccess extends MethodAccess {
+  GetInitParameterAccess() { this.getMethod() instanceof GetInitParameter }
+}
+
+/* Init parameter input of a Java EE web application. */
+class InitParameterInput extends LocalUserInput {
+  InitParameterInput() { this.asExpr() instanceof GetInitParameterAccess }
+}
+
+private class LessThanSanitizer extends DataFlow::BarrierGuard {
+  LessThanSanitizer() { this instanceof LTExpr }
+
+  override predicate checks(Expr e, boolean branch) {
+    e = this.(LTExpr).getLeftOperand() and
+    branch = true
+  }
+}
+
+/** Taint configuration of uncontrolled thread resource consumption. */
+class ThreadResourceAbuse extends TaintTracking::Configuration {
+  ThreadResourceAbuse() { this = "ThreadResourceAbuse" }
+
+  override predicate isSource(DataFlow::Node source) {
+    source instanceof RemoteFlowSource or source instanceof LocalUserInput
+  }
+
+  override predicate isSink(DataFlow::Node sink) { sink instanceof PauseThreadSink }
+
+  override predicate isAdditionalTaintStep(DataFlow::Node node1, DataFlow::Node node2) {
+    exists(ConditionalExpr ce | ce.getAChildExpr() = node1.asExpr() and ce = node2.asExpr()) // request.getParameter("nodelay") != null ? 0 : sleepTime
+    or
+    exists(
+      Method rm, ClassInstanceExpr ce, Argument arg, FieldAccess fa // thread.start() invokes the run() method of thread implementation
+    |
+      rm.hasName("run") and
+      ce.getConstructedType() = rm.getSourceDeclaration().getDeclaringType() and
+      ce.getConstructedType().getASupertype*().hasQualifiedName("java.lang", "Runnable") and
+      ce.getAnArgument() = arg and
+      fa = rm.getAnAccessedField().getAnAccess() and
+      arg.getType() = fa.getField().getType() and
```

Comment:
Maybe this could instead check if the respective parameter is (using localFlow?) assigned to the field; to reduce some false positives.

Reply:
@Marcono1234 - there is no localFlow like the following
```
DataFlow::localExprFlow(arg, fa)
```
So we cannot use such a flow to reduce FPs. Is there a way to achieve the desired flow using localFlow? Please shed some light on this. Thanks.

Reply:
Not sure if there is something built-in, but you could
1. Get the index of the argument
2. Get the corresponding constructor parameter at that index (might not work correctly for varargs, but can probably be ignored here)
3. Check for local flow from the constructor parameter to `fa.getField().getAnAssignedValue()`

Though I don't think CodeQL has any predicates for local taint flow, so hopefully the proposed change does not cause too many false negatives.

Or maybe I misunderstood your intention with this predicate.

Reply:
Hi @Marcono1234,

Thanks for the advice but I don't think it's my intention with the predicate. The scenario I want to model is:
- A runnable implementation with a constructor that passes the parameter of sleep time e.g. <code>delayTime</code> in the test case:
```new UncheckedSyncAction(delayTime)```
- Value of the <code>delayTime</code> parameter is set to an instance variable e.g. <code>waitTime</code>
- Invoke the <code>start()</code> method on the runnable will invoke the <code>run</code> method of the runnable without any argument
- In the <code>run</code>method, access the <code>waitTime</code> instance variable to set thread sleep time

Reply:
> Value of the `delayTime` parameter is set to an instance variable e.g. `waitTime`

That is what my comment above suggested: local flow from the parameter to `fa.getField().getAnAssignedValue()`.
The rest (`start()` method and how `run()` then uses the value) would not be affected, since that is covered by different parts of your query.

Reply:
Thanks @Marcono1234 for the clarification. I've fixed the query. Now it's much more precise and cleaner.